internal/provider/resources/resource_api_token.go

```diff
@@ -0,0 +1,547 @@
+
+	// Save updated data into Terraform state
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *ApiTokenResource) Delete(
+	ctx context.Context,
+	req resource.DeleteRequest,
+	resp *resource.DeleteResponse,
+) {
+	var data models.ApiTokenResource
+
+	// Read Terraform prior state data into the model
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// delete request
+	apiToken, err := r.IamClient.DeleteApiTokenWithResponse(
+		ctx,
+		r.OrganizationId,
+		data.Id.ValueString(),
+	)
+	if err != nil {
+		tflog.Error(ctx, "failed to delete API token", map[string]interface{}{"error": err})
+		resp.Diagnostics.AddError(
+			"Client Error",
+			fmt.Sprintf("Unable to delete API token, got error: %s", err),
+		)
+		return
+	}
+	statusCode, diagnostic := clients.NormalizeAPIError(ctx, apiToken.HTTPResponse, apiToken.Body)
+	// It is recommended to ignore 404 Resource Not Found errors when deleting a resource
+	if statusCode != http.StatusNotFound && diagnostic != nil {
+		resp.Diagnostics.Append(diagnostic)
+		return
+	}
+
+	tflog.Trace(ctx, fmt.Sprintf("deleted an API token resource: %v", data.Id.ValueString()))
+}
+
+func (r *ApiTokenResource) ImportState(
+	ctx context.Context,
+	req resource.ImportStateRequest,
+	resp *resource.ImportStateResponse,
+) {
+	resource.ImportStatePassthroughID(ctx, path.Root("id"), req, resp)
+}
+
+func (r *ApiTokenResource) ValidateConfig(
+	ctx context.Context,
+	req resource.ValidateConfigRequest,
+	resp *resource.ValidateConfigResponse,
+) {
+	var data models.ApiTokenResource
+
+	resp.Diagnostics.Append(req.Config.Get(ctx, &data)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Convert Terraform set of roles to API token roles
+	roles, diags := RequestApiTokenRoles(ctx, data.Roles)
+	if diags.HasError() {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	tokenRole, diags := RequestApiTokenPrimaryRole(roles, data.Type.ValueString())
+	if diags != nil {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+
+	entityType := data.Type.ValueString()
+
+	// Check if the role is valid for the token entity type
+	if !utils.ValidateRoleMatchesEntityType(tokenRole.Role, entityType) {
+		resp.Diagnostics.AddError(
+			fmt.Sprintf("Role '%s' is not valid for token type '%s'", tokenRole.Role, entityType),
+			fmt.Sprintf("Please provide a valid role for the entity type '%s'", entityType),
+		)
+		return
+	}
+
+	diags = validateApiTokenRoles(entityType, roles)
+	if diags != nil {
+		resp.Diagnostics.Append(diags...)
+		return
+	}
+}
+
+func validateApiTokenRoles(entityType string, roles []iam.ApiTokenRole) diag.Diagnostics {
+	var numRolesMatchingEntityType int
+	var invalidRoleError string
+
+	for _, role := range roles {
+		if entityType == string(iam.ApiTokenRoleEntityTypeWORKSPACE) && role.EntityType == iam.ApiTokenRoleEntityTypeORGANIZATION {
```

Comment:
can we check that if `entityType` is organization then the `organizationId` matches `r.OrganizationId` and add a test case as well for when it doesnt

Reply:
we can't include this validation because `r.OrganizationId` is not set when `ValidateConfig` is run

Reply:
can we add that as a comment in the code as well